### PR TITLE
Pause sync functionality

### PR DESCRIPTION
**Pull request by cooktheryan**
Signed-off-by: Ryan Cook <rcook@redhat.com>

**Describe what this PR does**
This PR introduces the flag PauseSync. Default value is false but it can be introduced and will delete the old job and replace it with a job that has parallelism set to 0. By not removing the destination it ensures the destination service IP will still be the same when source comes back online.

**Is there anything that requires special attention?**
This is my first time writing GO :fearful: 

**Related issues:**
#35

### CODE REVIEW

File: controllers/replicationdestination_controller.go [hunk 1]
```diff
@@ -288,6 +288,14 @@ func (r *rsyncDestReconciler) ensureJob(l logr.Logger) (bool, error) {
 		}
 		backoffLimit := int32(2)
 		r.job.Spec.BackoffLimit = &backoffLimit
+		if !r.Instance.Spec.Rsync.PauseSync {
```
Comment (screeley44) [inline controllers/replicationdestination_controller.go:291]: should this be an if...else block? you only save one line of code, not sure it really matters other than readability and style...

**Comment by screeley44** [comment]
I think you need to regenerate the deepcopy autogenerated file since you modified the api types

File: controllers/replicationdestination_controller.go (hunk 1 above)
Comment (cooktheryan) [reply controllers/replicationdestination_controller.go:291]: I swear I have written code before :facepalm:

File: api/v1alpha1/replicationdestination_types.go [hunk 2]
```diff
@@ -102,6 +102,9 @@ type ReplicationDestinationRsyncSpec struct {
 	// sshUser is the username for outgoing SSH connections. Defaults to "root".
 	//+optional
 	SSHUser *string `json:"sshUser,omitempty"`
+	// pauseSync bool to pause the rsync of data". Defaults to false.
+	//+optional
+	PauseSync bool `json:"pauseSync,omitempty"`
```
Comment (JohnStrunk) [inline api/v1alpha1/replicationdestination_types.go:107]: The goal is typically "declarative" phrasing... perhaps `paused` or `disabled` for the field name.

Also, the struct comment is what is output when a user does a `kubectl explain ...`. How about something like "XXX can be used to temporarily stop data replication. Defaults to false".

File: controllers/replicationdestination_controller.go [hunk 3]
```diff
@@ -334,6 +341,13 @@ func (r *rsyncDestReconciler) ensureJob(l logr.Logger) (bool, error) {
 		return false, err
 	}
 
+	// If PauseSync has been issued but Parallelism is not 0 delete job
+	if r.Instance.Spec.Rsync.PauseSync && *r.job.Spec.Parallelism != int32(0) {
```
Comment (JohnStrunk) [inline controllers/replicationdestination_controller.go:345]: I don't know if `parallelism` is mutable...
If it is mutable, this shouldn't be necessary.
If not, you're going to run into problems w/ `CreateOrUpdate`, above. You'll need to put the parallelism stuff inside a `if r.job.CreationTimestamp.IsZero() {` check so you only try to set it on 1st create. (See PVC handling code in `volumehandler.go`)

Also, this handles the case where you "set" pause. What about "unsetting" it? (pause==false; parallelism==0)

**Review by JohnStrunk — CHANGES_REQUESTED** [review]
Some general thoughts:
* The source doesn't get any indication that replication is paused, so it will continue to retry w/ no real indication of why it repeatedly fails. I don't really know how to easily fix that... something for the future
* Should we be able to pause the source as well?

File: controllers/replicationdestination_controller.go [hunk 4]
```diff
@@ -315,7 +315,17 @@ func (r *rsyncDestReconciler) ensureJob(l logr.Logger) (bool, error) {
 			r.job.Spec.Template.ObjectMeta.Labels[k] = v
 		}
 		backoffLimit := int32(2)
+		disabledSync := "disabled"
+		enabledSync := "enabled"
+		syncValue := *r.Instance.Spec.Rsync.DataSync
```
Comment (cooktheryan) [inline controllers/replicationdestination_controller.go:320]: @JohnStrunk any ideas here? 

When this variable is defined it causes the tests to fail.

**Comment by cooktheryan** [comment]
@JohnStrunk I can roll back the const's in the controllers if it makes no sense. When running make test it suggested I switch them to constants because they were defined multiple times.

I also took a stab at some test cases because I don't believe I was going to be able to merge without messing up code coverage

**Comment by cooktheryan** [comment]
@JohnStrunk the items work as expected. I can't seem to create a clean set of test cases but the functionality works and the job doesn't actually delete it just gets updated.

**Pull request closed without merging**